File: src/net/rtp/sender.cpp
#include "sender.hpp"

#include "packet.hpp"
#include "time.hpp"


namespace shar::net::rtp {

static const u16 MTU = 1000;

PacketSender::PacketSender(Context context, IpAddress ip, Port port)
    : Context(std::move(context))
    , m_endpoint(std::move(ip), port)
    , m_context()
    , m_socket(m_context)
    , m_packetizer(MTU)
    , m_sequence(0)
    , m_bytes_sent(0)
    , m_fragments_sent(0)
    , m_client(m_context)
    {}

void PacketSender::connect() {
  // 1. Connect to shar server
  // 2. Create session
  // 3. Wait for connection notification
  // 4. ICE
  //
  // set ip = client ip
  // and monitor connectivity
  m_client.connect(
    IpAddress::from_string("127.0.0.1"), 1337,
    [](ErrorCode ec) {
      if (!ec) {
        LOG_INFO("Successfully connected to ICE server");
      }
      else {
        LOG_ERROR("Failed to connect to ICE server: {}", ec.message());
      }

      // TODO: push continuation task on m_context
    }
  );

  m_context.run_for(Seconds(3));
  if (!m_client.connected()) {
    // TODO: retry
    // TODO: notify GUI thread about connection issues
    throw std::runtime_error("Connection to ICE server timed out");
  }

  bool created = false;
  m_client.open_session(
    "kekus",
    [&created](ice::SessionID session) {
      LOG_INFO("Session created. ID: {}", session);
      created = true;
    },
    [](ice::ConnectionInfo info) {
      LOG_INFO("Client connection notification received: {}", info);

      // FIXME: gather candidates and serialize
      return ice::ConnectionInfo{};
    }
    );
  m_context.run_for(Seconds(3));

  if (!created) {
    throw std::runtime_error("CreateSession request timed out");
  }
}

void PacketSender::run(Receiver<Unit> packets) {
  connect();

  m_socket.open(udp::v4());

  // TODO: use port range instead of constant
  auto ip = IpAddress(IPv4({ 0, 0, 0, 0 }));
  auto endpoint = udp::Endpoint(ip, Port{44444});
  m_socket.bind(endpoint);

  auto sent = Metric(m_metrics, "bytes sent", Metrics::Format::Bytes);
  while (auto packet = packets.receive()) {
    if (m_running.expired()) {
      break;
    }

    sent += packet->size();
    set_packet(std::move(*packet));
    send();
  }

  shutdown();
  m_socket.close();
}

void PacketSender::shutdown() {
    m_running.cancel();
}

void PacketSender::set_packet(Unit packet) {
    m_current_packet = std::move(packet);
    m_packetizer.set(m_current_packet.data(), m_current_packet.size());

    m_bytes_sent = 0;
}

void PacketSender::send() {
  static const usize HEADER_SIZE = rtp::Packet::MIN_SIZE;
  std::array<u8, MTU + HEADER_SIZE> buffer;

  while (auto fragment = m_packetizer.next()) {
    // sleep for 1ms every 128 fragments
    if ((++m_fragments_sent & (128 - 1)) == 0) {
      std::this_thread::sleep_for(Milliseconds(1));
    }

    assert(buffer.size() >= fragment.size() + HEADER_SIZE);

    // setup packet
    std::memset(buffer.data(), 0, HEADER_SIZE);
    std::memcpy(buffer.data() + HEADER_SIZE, fragment.data(), fragment.size());

    rtp::Packet packet(buffer.data(), HEADER_SIZE + fragment.size());
    packet.set_version(2);
    packet.set_has_padding(false);
    packet.set_has_extensions(false);
    packet.set_contributors_count(0);
    packet.set_marked(false);
    packet.set_payload_type(96);
    packet.set_sequence(m_sequence++);
    packet.set_timestamp(m_current_packet.timestamp());
    packet.set_stream_id(0);

    ErrorCode ec;
    m_socket.send_to(span(packet), m_endpoint, 0, ec);

    if (ec) {
      LOG_ERROR("Failed to send rtp packet: {}", ec.message());
    }

  }
}

}

File: src/net/rtp/sender.rs
use std::io;
use std::net::SocketAddr;
use std::time::Duration;

use futures::stream::{Stream, StreamExt};
use tokio::net::UdpSocket;

use super::packet::Packet;
use super::packetizer::Packetizer;
use crate::codec::Unit;

pub struct Sender<S> {
    units: S,

    sequence: u16,
}

impl<S, U> Sender<S>
where
    S: Stream<Item = U> + Unpin,
    U: Unit,
{
    pub fn new(units: S) -> Self {
        Sender { units, sequence: 0 }
    }

    pub async fn stream_to(&mut self, remote: SocketAddr) -> io::Result<()> {
        let local: SocketAddr = ([0, 0, 0, 0], 0).into();
        // ...why the heck are these async? -_-
        let mut socket = UdpSocket::bind(local).await?;
        socket.connect(remote).await?;

        let mut npackets = 0usize;

        let mut buffer = [0u8; 2048];
        while let Some(unit) = self.units.next().await {
            let data = unit.data();
            let mtu = 1400;
            let mut packetizer = Packetizer::with_mtu(mtu, data);

            while let Some(fragment) = packetizer.next() {
                const HEADER_SIZE: usize = 12;
                let packet_size = HEADER_SIZE + 2 + fragment.payload.len();
                let mut packet = Packet::new(&mut buffer[..packet_size]).unwrap();

                let header = packet.header_mut();
                header.set_version(2);
                header.set_payload_type(96);
                header.set_sequence(self.sequence);
                header.set_timestamp(unit.timestamp() as u32);

                let payload = packet.payload_mut();
                payload[0] = fragment.indicator;
                payload[1] = fragment.header;
                payload[2..].copy_from_slice(fragment.payload);

                self.sequence = self.sequence.wrapping_add(1);

                socket.send(&buffer[..packet_size]).await?;

                npackets += 1;
                if npackets % 128 == 0 {
                    tokio::time::delay_for(Duration::from_millis(1)).await;
                }
            }
        }

        Ok(())
    }
}
